ldap/servers/slapd/daemon.c

```diff
@@ -2030,6 +2262,48 @@ handle_new_connection(Connection_Table *ct, int tcps, PRFileDesc *listenfd, int
         connection_table_move_connection_on_to_active_list(the_connection_table, conn);
     }
 
+#ifdef ENABLE_EPOLL
+    /* Set up the epoll event for this connection */
+    conn->c_event->events = EPOLL_EVENTS;
+    conn->c_event->data.ptr = conn;
+    slapi_log_err(SLAPI_LOG_DEBUG, "handle_new_connection",
+                  "Adding connection %p (descriptor %d, table %d, conn %d) to epoll_fd %d with flags %s\n",
+                  conn, PR_FileDesc2NativeHandle(conn->c_prfd), conn->c_ct_list, conn->c_ci,
+                  the_connection_table->epoll_fd[conn->c_ct_list], epoll_event_flags_to_string(conn->c_event->events));
+
+    /* Add the connection to the epoll instance */
+    if (epoll_ctl(the_connection_table->epoll_fd[conn->c_ct_list], EPOLL_CTL_ADD, PR_FileDesc2NativeHandle(conn->c_prfd), conn->c_event) == -1) {
+        slapi_log_err(SLAPI_LOG_ERR, "handle_new_connection", "epoll_ctl() failed: %s\n",
+                      strerror(errno));
+        /* XXX cleanup how? */
+    }
+
+    if ((conn->c_idle_tfd = timerfd_create(CLOCK_MONOTONIC, TFD_NONBLOCK)) < 0) {
+        slapi_log_err(SLAPI_LOG_ERR, "handle_new_connection", "timerfd_create() failed: %s\n",
+                      strerror(errno));
+        /* XXX cleanup how? */
```

Comment:
If we can't create this, or add to epoll, I think we go straight into disconnect_server_nomutex() so that we can disconnect the connection and log an appropriate error. Generally both of these will error if we are out of fd's or some other resource limit so I think disconnecting and making noise is correct here.

Reply:
I think you're right. Up until the connection is added to the active list, it seems there is a bit of an odd limbo state here where we don't have a clear way to return this to the freelist on the conntable. I suspect we need to break `connection_table_move_connection_out_of_active_list`. Maybe the code after https://github.com/389ds/389-ds-base/blob/main/ldap/servers/slapd/conntable.c#L467 needs to become a separate function where we can return the connection from these early error cases where disconnect server mutex isn't the right answer, and before we are on the active list.

Really, the active list and the concept of it isn't needed once we have epoll, but that's a bigger cleanup and rework.